### ISSUE TITLE
Type refining Redux actions without switch statements

### ISSUE THREAD

**Issue by jjshammas** [facebook/flow#5396]
A common pattern in Redux is to type a Redux action as a union of all possible action shapes:

```javascript
type Action =
    | { type: "A", foo: "bar" }
    | { type: "B" }
	;
```

Then, as long as you are using an if or a switch statement, Flow can infer which action was passed into a function (a reducer):

```javascript
function reducer(action: Action) {
  if (action.type === "A") {
    console.log(action.foo.length); // no error here!
  }
}
```

However, a common pattern to avoid the boilerplate of adding dozens of conditionals or case statements is to create a map of handlers instead. This doesn't play nice with Flow:

```javascript
const actionHandlers = {
  "A": (action) => console.log(action.foo.length), // throws an error accessing foo
  "B": (action) => {},
}

function reducer(action: Action) {
  return actionHandlers.hasOwnProperty(action.type)
  	? actionHandlers[action.type](action)
  	: {};
}
```

This significantly limits how a reducer can be defined or generated, since without a top-level set of ifs or cases, Flow doesn't understand how to refine the `Action` type into a more specific type.

Is there a best practice for using objects to pick an action handler while keeping the reducer Flow-compliant?

**Comment by CharlesStover** [comment]
You probably have to actually define the types of actions that each action handler will be receiving. Since you are just saying that it receives an action, Flow expects it to be able to receive _any_ action. Even though you code is not currently sending it those actions, it doesn't know that it won't be sent those actions in the future.

```JS

type Action = ActionA | ActionB | ActionC;

const actionHandlers = {
  "A": (action: ActionA) => console.log(action.foo.length), // throws an error accessing foo
  "B": (action: ActionB) => {},
}
```

**Comment by jjshammas** [comment]
@CharlesStover Unfortunately, that's also an error. It fixes my original error from being thrown inside `actionHandlers`, but then the issue is that the `reducer` has just an `Action` type object, and is trying to send it into a function expecting `ActionA`.

The core issue is that Flow seemingly cannot type refine without an if or switch statement. I'm really surprised there's not a known solution (that I could find) since Redux themselves recommend this object pattern: https://redux.js.org/docs/recipes/ReducingBoilerplate.html#generating-reducers

**Comment by shinster** [comment]
If you define your actions in an object keyed by their type property, something like this can work with Flow:
```
type ActionsObj = ValidatedActions<{
  foo: {type: 'foo', qux: string, baz: boolean},
  bar: {type: 'bar', qux: number},
}>

type ValidatedActions<O> = $ObjMap<O, <A>(A) => A & $ElementType<O, $ElementType<A, 'type'>>>

type Action = $Values<ActionsObj>

type State = {name: string, count: number}

type Handlers<O> = $ObjMap<O, <A>(A) => (State, A) => State>

const handlers: Handlers<ActionsObj> = {
  foo: (state, action) => ({...state, name: action.baz ? action.qux : 'abc'}),
  bar: (state, action) => ({...state, count: state.count + action.qux})
}

const initialState = {name: '', count: 0}
        
const reducer = (state?: State = initialState, action: Action): State => (
  handlers.hasOwnProperty(action.type)
    ? handlers[action.type](state, action)
    : state
)
```
There's a little redundancy from having to repeat the type prop of each action, but at least Flow can enforce that the keys and action types are equal for each entry.

**Comment by jjshammas** [comment]
@shinster thanks for your suggestion—this seems like a really valuable solution, though it only works for small projects. The `handlers` object must have a key for every possible type, which makes it difficult to split up handler logic into multiple `handlers`.

Similarly, splitting up ActionsObj causes some weirdness:

```
type UserActions = ValidatedActions<{
  foo: {type: 'foo', qux: string, baz: boolean},
  bar: {type: 'bar', qux: number},
}>;

type SettingsActions = ValidatedActions<{
  qux: {type: "qux", payload: string},
}>;

type AllActions = UserActions & SettingsActions;
```

For some reason, Flow seems to ignore `SettingsActions` and doesn't give any errors related to its action definitions.

**Comment by shinster** [comment]
I've found that when combining object types, sometimes the spread operator gives better results than intersections. The objects that are spread should be exact types though.

Here's a version with the reducers split up:
````
type UserActions = ValidatedActions<{
  foo: {type: 'foo', qux: string, baz: boolean},
  bar: {type: 'bar', qux: number},
}>

type SettingsActions = ValidatedActions<{
  qux: {type: "qux", payload: string},
}>

type AllActions = {|
  ...UserActions,
  ...SettingsActions,
|}

type ValidatedActions<Actions> = $Exact<$ObjMap<Actions,
  <A>(A) => {type: $PropertyType<A, 'type'>} & $ElementType<Actions, $PropertyType<A, 'type'>>
>>

type UserState = {name: string, count: number}
type SettingsState = {test: string}

type Handlers<State, Actions> = $Exact<$ObjMap<Actions, <A>(A) => (State, A) => State>>

const userHandlers: Handlers<UserState, UserActions> = {
  foo: (state, action) => ({...state, name: action.baz ? action.qux : 'abc'}),
  bar: (state, action) => ({...state, count: state.count + action.qux}),
}

const settingsHandlers: Handlers<SettingsState, SettingsActions> = {
  qux: (state, action) => ({...state, test: state.test + action.payload}),
}

const initialUser: UserState = {name: '', count: 0}
const initialSettings: SettingsState = {test: 'aaa'}

const makeReducer = <S, H: Handlers<S, *>>(initialState: S, handlers: H) =>
  (state?: S = initialState, action: $Values<AllActions>): S =>
    handlers.hasOwnProperty(action.type)
      ? handlers[action.type](state, action)
      : state

const userReducer = makeReducer(initialUser, userHandlers)
const settingsReducer = makeReducer(initialSettings, settingsHandlers)
````

**Comment by jjshammas** [comment]
Thanks for this, @shinster. Really good stuff. I made a small change of making the `Handlers` support null for any value:

`type Handlers<State, Actions> = $Exact<$ObjMap<Actions, <A>(A) => ?(State, A) => State>>`

And created a, simply named, `Action` type for use in annotating the `dispatch` function and others:

`type Action = $Values<AllActions>`

Unfortunately, at this step this approach failed for our use case. Flow is very slow with large unions (#5631), so I was hoping this approach using maps would be far faster. It *was* much faster, until we annotated the `dispatch` function, at which point Flow completely failed because it timed out trying to merge together a large number of actions and compare to it every time the dispatch function was called.

Statically typing redux in large environments continues to be quite difficult. Perhaps the only option is leave methods like dispatch ambiguous—only requiring that an object be passed in—though you lose a tremendous amount of valuable type checking.
